File: src/main/java/de/maxhenkel/pipez/utils/PipeEnergyStorage.java
package de.maxhenkel.pipez.utils;

import de.maxhenkel.pipez.blocks.tileentity.PipeLogicTileEntity;
import de.maxhenkel.pipez.blocks.tileentity.types.EnergyPipeType;
import net.minecraft.core.Direction;
import net.neoforged.neoforge.transfer.energy.EnergyHandler;
import net.neoforged.neoforge.transfer.transaction.TransactionContext;

public class PipeEnergyStorage implements EnergyHandler {

    protected PipeLogicTileEntity pipe;
    protected Direction side;
    protected long lastReceived;

    public PipeEnergyStorage(PipeLogicTileEntity pipe, Direction side) {
        this.pipe = pipe;
        this.side = side;
    }

    public void tick() {
        if (pipe.getLevel().getGameTime() - lastReceived > 1) {
            EnergyPipeType.INSTANCE.pullEnergy(pipe, side);
        }
    }

    @Override
    public long getAmountAsLong() {
        return 0;
    }

    @Override
    public long getCapacityAsLong() {
        return Long.MAX_VALUE;
    }

    @Override
    public int insert(int amount, TransactionContext transaction) {
        //TODO Check if this needs to be rolled back
        lastReceived = pipe.getLevel().getGameTime();
        return EnergyPipeType.INSTANCE.receive(pipe, side, amount, transaction);
    }

    @Override
    public int extract(int amount, TransactionContext transaction) {
        return 0;
    }
}

File: src/main/java/de/maxhenkel/pipez/blocks/tileentity/storage/PipeEnergyStorage.kt
package de.maxhenkel.pipez.blocks.tileentity.storage

import de.maxhenkel.pipez.blocks.tileentity.PipeLogicTileEntity
import de.maxhenkel.pipez.blocks.tileentity.types.EnergyPipeType
import de.maxhenkel.pipez.blocks.tileentity.types.FluidPipeType
import de.maxhenkel.pipez.types.EnergyMaterial
import de.maxhenkel.pipez.types.PipeSide
import net.minecraft.core.Direction
import net.minecraftforge.energy.IEnergyStorage
import kotlin.math.abs

class PipeEnergyStorage(pipe: PipeLogicTileEntity, side: Direction) : PullPushStorage(pipe, side), IEnergyStorage {
    override fun tick() {
        super.tick()
    }

    override fun pull() {
        EnergyPipeType.INSTANCE.ejectMaterial(pipe, side)
    }

    override fun receiveEnergy(maxReceive: Int, simulate: Boolean): Int {
        updatePushed()
        return EnergyPipeType.INSTANCE.receiveMaterial(
            pipe, PipeSide.fromDirection(side), EnergyMaterial(maxReceive), simulate).toInt()
    }

    override fun extractEnergy(maxExtract: Int, simulate: Boolean): Int {
        return 0
    }

    override fun getEnergyStored(): Int {
        return 0
    }

    override fun getMaxEnergyStored(): Int {
        return Int.MAX_VALUE
    }

    override fun canExtract(): Boolean {
        return false
    }

    override fun canReceive(): Boolean {
        return true
    }
}